scripts/opentitan/autoreg.py

```diff
@@ -900,36 +968,163 @@ def _generate_type(self, tfp: TextIO) -> None:
+                bitfields[radix] = reg.fields
+            elif radix in shared_fields:
+                shared_fields[radix][1].add(reg.name)
+            else:
+                flds = reg.fields
+                if flds and (len(flds) > 1 or flds[0].width != self._regwidth):
+                    bitfields[reg.name] = reg.fields
+        reg = None
+        for reg in self._registers:
+            self._generate_register(reg, shared_fields, nibcount,
+                                         reg.count, tfp)
+        address = reg.address + self._regwidth // 8 if reg else 0
+        address = reg.address + (reg.count * self._regwidth) // 8 if reg else 0
+        print(f'        (0x{address:0{nibcount}x} => @END),', file=tfp)
+        print('    }\n}\n', file=tfp)
+        print(f'register_bitfields! [u{self._regwidth},', file=tfp)
+        for name, fields in bitfields.items():
+            print(f'    {name} [', file=tfp)
+            for fld in fields:
+                print(f'        {fld.name} OFFSET({fld.offset}) '
+                      f'NUMBITS({fld.width}) [],', file=tfp)
+            print('    ],', file=tfp)
+        print('];\n', file=tfp)
+
+    def _generate_register(self, reg: AutoRegister,
+            shfields: dict[str, tuple[list[AutoField], set[str]]],
+            nibcount: int, repcount:int, tfp: TextIO) -> None:
+        regtype = {
+            AutoAccess.RO: 'ReadOnly',
+            AutoAccess.WO: 'WriteOnly',
+        }.get(reg.access, 'ReadWrite')
+        regdesc = ''
+        radix = reg.name.rsplit('_', 1)[0]
+        if radix in shfields and reg.name in shfields[radix][1]:
+            fields = shfields[radix][0]
+            regdescname = radix
+        else:
+            fields = reg.fields
+            regdescname = reg.name
+        if fields and (len(fields) > 1 or fields[0].width != self._regwidth):
+            regdesc = f', {regdescname.upper()}::Register'
+        if repcount > 1:
+            print(f'        (0x{reg.address:0{nibcount}x} => '
+                  f'{reg.name.lower()}: '
+                  f'[{regtype}<u{self._regwidth}{regdesc}>; {repcount}]),',
+                  file=tfp)
+        else:
+            print(f'        (0x{reg.address:0{nibcount}x} => '
+                  f'{reg.name.lower()}: '
+                  f'{regtype}<u{self._regwidth}{regdesc}>),', file=tfp)
+        self._log.warning('Reg: %s', reg)
 
 
 def main():
     """Main routine"""
     debug = True
-    desc = sys.modules[__name__].__doc__.split('.', 1)[0].strip()
+    mod = sys.modules[__name__]
+    desc = mod.__doc__.split('.', 1)[0].strip()
     argparser = ArgumentParser(description=f'{desc}.')
+    base = 'AutoReg'
+    autoregs = {
+        name.removesuffix(base).lower(): tit
+        for name, tit in [(m, getattr(mod, m)) for m in dir(mod)]
+        if isinstance(tit, type) and issubclass(tit, AutoReg) and name != base
+    }
+    outkinds = sorted(autoregs, key=lambda n: '' if n == 'qemu' else n)
```

Comment:
This is also repeated in `QEMUAutoTop.outkind`, but maybe you could name this 'qemu' string to make what is going on here a bit clearer? If I'm understanding correctly, the intention is essentially:

```suggestion
    DEFAULT_OUTKIND = 'qemu'
    outkinds = sorted(autoregs, key=lambda n: '' if n == DEFAULT_OUTKIND else n)
```

If you want, you could even extract this to some `sort_outkinds` function which does this for you, rather than repeating the logic.

Reply:
Done in both scripts.